### PR TITLE
Add mariadb 10.3 + mysql 8 to Drone, remove 10.2 workarounds

### PR DESCRIPTION
**Pull request by PVince81**
Add mariadb 10.3 to the unit test matrix.

For https://github.com/owncloud/core/issues/29483 but doesn't cover mb4 AFAIK

### CODE REVIEW

**Comment by PVince81** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by DeepDiver1975** [comment]
results look ugly ... needs debugging ....

**Comment by VicDeo** [comment]
@DeepDiver1975 not ugly but predictable. `mb4` support is detected basing on `innodb_file_format == barracuda` check. `innodb_file_format` option was deprecated in mariadb 10.2 and removed in mariadb 10.3
see https://github.com/owncloud/core/issues/29483#issuecomment-378335745 if you need more details

**Comment by DeepDiver1975** [comment]
I was missing the whole conversation - THX

Can you add this maria db version check? THX

**Comment by PVince81** [comment]
the [innodb_file_format was deprecated then removed](https://mariadb.com/kb/en/library/xtradbinnodb-server-system-variables/#innodb_file_format) in 10.3.

while we could rely on its value to default to Barracuda for 10.2.2 there is still the likeliness of distribution packages having the other value in their config.

I think we can first check if these values exist and rely on them. If the variables are gone, fall back to a DB version check. I see the DB version appear in some variables:
```
MariaDB [(none)]> show variables like '%ver%';
+------------------------------------------+----------------------------------+
| Variable_name                            | Value                            |
+------------------------------------------+----------------------------------+
| aria_force_start_after_recovery_failures | 0                                |
| aria_recover_options                     | BACKUP,QUICK                     |
| character_set_server                     | utf8mb4                          |
| collation_server                         | utf8mb4_general_ci               |
| innodb_force_recovery                    | 0                                |
| innodb_ft_server_stopword_table          |                                  |
| innodb_show_verbose_locks                | 0                                |
| innodb_version                           | 5.7.23                           |
| log_slow_verbosity                       |                                  |
| master_verify_checksum                   | OFF                              |
| myisam_recover_options                   | BACKUP,QUICK                     |
| protocol_version                         | 10                               |
| relay_log_recovery                       | OFF                              |
| server_id                                | 1                                |
| slave_sql_verify_checksum                | ON                               |
| slave_type_conversions                   |                                  |
| thread_pool_oversubscribe                | 3                                |
| version                                  | 10.2.18-MariaDB                  |
| version_comment                          | openSUSE package                 |
| version_compile_machine                  | x86_64                           |
| version_compile_os                       | Linux                            |
| version_malloc_library                   | system                           |
| version_ssl_library                      | OpenSSL 1.1.0h-fips  27 Mar 2018 |
| wsrep_convert_lock_to_trx                | OFF                              |
| wsrep_patch_version                      | wsrep_25.23                      |
| wsrep_recover                            | OFF                              |
+------------------------------------------+----------------------------------+
```

Interestingly there's also a value character_set_server and collation_server.

This is what I see on my local machine for 10.2:
```
MariaDB [(none)]> show variables like '%character_set%';
+--------------------------+------------------------------+
| Variable_name            | Value                        |
+--------------------------+------------------------------+
| character_set_client     | utf8                         |
| character_set_connection | utf8                         |
| character_set_database   | utf8mb4                      |
| character_set_filesystem | binary                       |
| character_set_results    | utf8                         |
| character_set_server     | utf8mb4                      |
| character_set_system     | utf8                         |
| character_sets_dir       | /usr/share/mariadb/charsets/ |
+--------------------------+------------------------------+
```

**Comment by VicDeo** [comment]
@PVince81 
```
PDO::getAttribute(PDO::ATTR_SERVER_VERSION);
```
and then borrow a code piece from Doctrine
https://github.com/doctrine/dbal/blob/90f79c8b39fa451264537440ead0fe1ab1c751b8/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php#L116-L119

**Comment by PVince81** [comment]
@VicDeo can I also kill https://github.com/owncloud/core/issues/29483#issuecomment-378335745 ? it was stated that we can remove this workaround once DBAL is updated. I need to check what DBAL version has this.

**Comment by PVince81** [comment]
sorry, meant https://github.com/owncloud/core/pull/29100 (edited comment now).
this was added as a workaround for something.

we're on dbal 2.8.0 on master

**Comment by PVince81** [comment]
reverted the removal of workaround, just in case...

**Comment by PVince81** [comment]
I still have no clue why this PR influences mysqlmb4 aka mysql 5.7. The logic should still be the same.

Also, I discovered this: https://dba.stackexchange.com/a/125893

With MySQL 8 the setting is also removed.

**Comment by PVince81** [comment]
ouch, of course... my logic is flawed :face_with_head_bandage:

**Comment by PVince81** [comment]
should be fixed now.

I've also added MySQL 8 in the matrix for fun and a added a check since MySQL 8 has removed the innodb vars as well.

**Comment by PVince81** [comment]
cool, all passed except MySQL 8 which has setup issues:
```
+ DB=mysql
+ install_cmd='maintenance:install -vvv       --database=mysql       --database-name=owncloud       --database-table-prefix=oc_       --admin-user=admin       --admin-pass=admin       --data-dir=/drone/src/data '
+ [[ mysql != \s\q\l\i\t\e ]]
+ install_cmd+=' --database-host=mysql                  --database-user=owncloud                  --database-pass=owncloud'
+ php occ maintenance:install -vvv --database=mysql --database-name=owncloud --database-table-prefix=oc_ --admin-user=admin --admin-pass=admin --data-dir=/drone/src/data --database-host=mysql --database-user=owncloud --database-pass=owncloud
Error while trying to create admin user: Failed to connect to the database: An exception occurred in driver: SQLSTATE[HY000] [2054] The server requested authentication method unknown to the client
->
```

need to check, maybe the default creds have changed for that docker

**Comment by PVince81** [comment]
related: http://php.net/manual/en/mysqli.requirements.php

I'll switch it to PHP 7.2 and we'll see

**Comment by PVince81** [comment]
doesn't work. maybe we comment out the mysql 8 test for now and adjust later ?

thoughts ? @DeepDiver1975 @VicDeo

**Comment by PVince81** [comment]
I re-read http://php.net/manual/en/mysqli.requirements.php and it says that one needs to change the my.cnf config of mysql. Not sure how to do that in this official container.

The PHP side support is said to be added in "future releases". So probably not worth the effort right now.

**Comment by PVince81** [comment]
fix: properly map mysql8 to actual mysql database

**Comment by PVince81** [comment]
```
+ wait-for-it mysql:3306
waiting for database to be ready
wait-for-it: waiting 15 seconds for mysql:3306
wait-for-it: timeout occurred after waiting 15 seconds for mysql:3306
```
for mysql8

hmm... where is that magic host name defined ?

**Comment by PVince81** [comment]
@DeepDiver1975 do you know where the host name "mysqlmb4" is being define ? is it part of the docker image ? I don't see the connection between the DB_TYPE and the host name that is used for waiting for connection.

**Comment by phil-davis** [comment]
Each service in the ``.drone.yml`` ``services:`` section has a name. When drone starts the service, it defines the name as a host name pointing to the drone IP address that the service is on.
So ``mysql8:`` is the magic that should make a host name ``mysql8`` valid.

**Comment by PVince81** [comment]
@phil-davis thank you, generous wizard. I have now applied your powerful magic to this pull request and hoping for the best.

**Comment by PVince81** [comment]
it worked !!!

@DeepDiver1975 @phil-davis @VicDeo please review

File: tests/drone/install-server.sh [hunk 1]
```diff
@@ -87,6 +87,10 @@ case "${DB_TYPE}" in
     wait-for-it mysqlmb4:3306
```
Comment (phil-davis) [inline tests/drone/install-server.sh]: Maybe this section can be deleted - not used any more?
Or maybe it can stay in case some dev cares?
No big deal

Comment (PVince81) [reply tests/drone/install-server.sh]: now that I understand the wizardry better, I'll remove this

**Review by phil-davis — COMMENTED** [review]
CI scripts LGTM.
I will let someone else comment on the "real" code changes.

**Comment by VicDeo** [comment]
@PVince81 Looks good. My quickfix for  MariaDB 10.2 is not needed any more as it is already supported as a Doctrine Platform in our current doctrine/dbal version.

How difficult it would be to write a test for `public function supports4ByteCharset(IDBConnection $connection) {`?

**Comment by VicDeo** [comment]
While considering a backport please note that stable10 still uses older `doctrine/dbal` and could not be updated to the release with direct MariaDB 10.2.6+ support due to PHP compatibility reasons.

**Comment by PVince81** [comment]
@VicDeo thanks for the warning. I don't think we should backport this unless MariaDB 10.2 becomes EOL soon. From what I see here https://mariadb.org/about/maintenance-policy/ MariaDB 10.2 still runs up to May 2022.

Now if all Linux distributions start shipping 10.3 by default we might need to backport anyway at some point.

**Comment by PVince81** [comment]
- [ ] add unit test for the utility method

**Comment by PVince81** [comment]
@VicDeo mocking database variables and query function ? challenge accepted: https://github.com/owncloud/core/pull/33255/commits/a2078b3440543770aa4f1f88dae03c82f5a69f85

**Comment by PVince81** [comment]
- [ ] discuss whether to backport

not sure how to read these results, seems that Ubuntu still is on MariaDB 10.2 here https://packages.ubuntu.com/search?keywords=mariadb-server

also see https://mariadb.com/kb/en/library/distributions-which-include-mariadb/

**Review by VicDeo — APPROVED** [review]
Great!

**Comment by phil-davis** [comment]
Backporting as part of "drop PHP 7.0" in `stable10` #35786 
Hopefully that will all be good and allow mariadb 10.3 support in ownCloud10

**Pull request merged**